beluga_tutorial/src/main.cpp

```diff
@@ -0,0 +1,278 @@
+    return 1;
+  }
+
+  const std::string directory = argv[1];
+  const std::string filename = argv[2];
+  const auto parameters_file_path = std::filesystem::path{directory} / filename;
+
+  TutorialParams tutorial_params;
+  load_params_from_yaml(parameters_file_path, tutorial_params);
+
+  // Create the TutorialDataset object for record the simulation data
+  TutorialDataset tutorial_dataset;
+
+  LandmarkMapVector landmark_map;
+  load_landmark_map_from_yaml(std::filesystem::path{directory} / tutorial_params.landmark_map_file_name, landmark_map);
+  tutorial_dataset.landmark_map = landmark_map;
+
+  // Generate particles
+  std::normal_distribution<double> initial_distribution(tutorial_params.initial_pose, tutorial_params.initial_pose_sd);
+  auto particles = beluga::views::sample(initial_distribution) |                       //
+                   ranges::views::transform(beluga::make_from_state<Particle>) |       //
+                   ranges::views::take_exactly(tutorial_params.number_of_particles) |  //
+                   ranges::to<beluga::TupleVector>;
+
+  // Execute the particle filter using beluga
+  double current_pose{tutorial_params.initial_pose};
+  for (auto n = 0; n < tutorial_params.number_of_cycles; ++n) {
+    TutorialData tutorial_data;
+
+    // Save ground truth
+    current_pose += tutorial_params.velocity * tutorial_params.dt;
+    tutorial_data.ground_truth = current_pose;
+
+    // Check if the simulation is out of bounds
+    if (current_pose > tutorial_params.map_size) {
+      break;
+    }
+
+    auto motion_model = [&](double state, auto& gen) {
+      const double distance = (tutorial_params.velocity * tutorial_params.dt);
+      std::normal_distribution<double> distribution(distance, tutorial_params.translation_sd);
+      return state + distribution(gen);
+    };
+
+    // Generate simulated sensor data
+    auto sensor_data =
+        landmark_map |                                                                                            //
+        ranges::views::transform([&](double landmark) { return landmark - current_pose; }) |                      //
+        ranges::views::remove_if([&](double range) { return std::abs(range) > tutorial_params.sensor_range; }) |  //
+        ranges::to<SensorData>;
+
+    auto sensor_model = [&](const double& state) {
+      auto particle_sensor_data = landmark_map |                                                           //
+                                  ranges::views::transform([&](const double lm) { return lm - state; }) |  //
+                                  ranges::to<SensorData>;
+
+      return tutorial_params.min_particle_weight +
+             std::transform_reduce(
+                 sensor_data.begin(), sensor_data.end(), 1.0, std::multiplies<>{}, [&](const auto& sensor_datum) {
+                   auto distances = particle_sensor_data |  //
+                                    ranges::views::transform([&](auto& particle_sensor_datum) {
```

Comment:
```suggestion
                                    ranges::views::transform([&](double particle_sensor_datum) {
```

Or pass by const-ref. Same everywhere.